[FEATURE] [API] Stocker des infos de l'utilisateur (claims de userInfo) lors d'une connexion SSO OIDC (PIX-9270)

lego-technix opened the pull request:
## :unicorn: Problème

Lors d'une connexion via un SSO OIDC, les _claims_ correspondant à la variable d'environnement `${IDP}_CLAIMS_TO_STORE` peuvent être définis (#7387) mais ne sont pas encore stockés.

## :robot: Proposition

Il y a 3 cas de figure pouvant se produire dans le cadre d'une authentification SSO OIDC : 

a. première connexion avec un compte SSO et création d'un nouveau compte Pix associé
b. première connexion avec un compte SSO et association avec un compte Pix préexistant
c. connexion avec un compte SSO déjà associé avec un compte Pix

La proposition est, **pour chacun de ces cas de figure**, de stocker les _claims_ correspondant à la variable d'environnement `${IDP}_CLAIMS_TO_STORE` dans le `authenticationComplement` de la table `authentication-methods` en écrasant toute potentielle valeur préexistante.

Les _claims_ définis par `${IDP}_CLAIMS_TO_STORE` sont des _claims_ **obligatoires**. Si un de ces _claims_ n’est pas récupéré à partir des _userInfo_, alors l’authentification échoue.

Les _claims_ spécifiés ne doivent pas obligatoirement être des _claims_ OIDC standards, même si c’est mieux. Pour info voici la liste des _claims_ OIDC standards : 
https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

On précise bien qu’un compte utilisateur ne peut avoir qu’une seule valeur pour un _claim_ donné. Par exemple un compte utilisateur ne peut avoir qu’un `employeeNumber`.

Le SSO Pôle emploi individu est pour l'instant exclu de ce dispositif.

## :rainbow: Remarques

### Correction d'un gros bug

Le  commit abdd328d027a43315daa8dc8d9d8d1367afde667 est un correctif d'un gros bug de copie par référence au lieu de copie par valeur (_copy by reference vs copy by value_) introduit par la PR #6229.

### Changement d'un null en undefined

Lorsqu'il n'y a pas d'`authenticationComplement` à stocker, la fonction `createAuthenticationComplement` renvoie maintenant `undefined` à la place de `null`, de manière à ce que dans ce cas la propriété `authenticationComplement` d'une `AuthenticationMethod` soit considérée comme absente et donc non à stocker, avec la règle _Joi_ associée : `Joi.object().instance(OidcAuthenticationComplement).empty()`

## :100: Pour tester

L'objectif est de tester les 3 cas de figure pouvant se produire dans le cadre d'une authentification SSO OIDC : 

a. première connexion avec un compte SSO et création d'un nouveau compte Pix associé
b. première connexion avec un compte SSO et association avec un compte Pix préexistant
c. connexion avec un compte SSO déjà associé avec un compte Pix

### Avec les RA

Au niveau des RA il n'est pas facile de tester les SSO, donc on ne teste pas dans cet environnement.

### En local 

1. Réinitialiser la BDD de manière à ce que plus aucun compte n'ait d'_authenticationMethod_ : 

   ```shell
   npm run db:reset
   ```

2. Définir la variable d'environnement suivante : 

   ```shell
   FWB_CLAIMS_TO_STORE=family_name, given_name, employeeNumber
   ```

3. Se connecter avec le SSO FWB avec n'importe quel compte du SSO

4. Au niveau de la double mire, choisir la partie _« Je n’ai pas de compte Pix »_ et cliquer sur le bouton _« Je crée mon compte »_

5. Après la création de compte réussie, vérifier en BDD que le compte créé a une `authentication-methods` : 

   ```sql
   select * from "authentication-methods" where "identityProvider"='FWB';
   ```

   avec un `authenticationComplement` qui contient une valeur JSON du type suivant : 
 
   ```json
   { "family_name": "AAAAA", "given_name": "BBBBB", "employeeNumber": "CCCCC" }
   ```

6. Se déconnecter.

7. Se connecter avec tous les autres SSO OIDC et vérifier qu'il n'y a pas de régression.

8. Vérifier que le `authenticationComplement` est absent pour toutes les `authentication-methods` des fournisseurs `CNAV`, `PAYSDELALOIRE` : 

   ```sql
   select * from "authentication-methods" where "identityProvider" in ('CNAV', 'PAYSDELALOIRE');
   ```

9. Vérifier que le `authenticationComplement` des `authentication-methods` du fournisseur `POLE_EMPLOI` n'est pas absent et contient toujours uniquement des données de connexion : 

   ```sql
   select * from "authentication-methods" where "identityProvider"='POLE_EMPLOI';
   ```
   avec un `authenticationComplement` qui contient une valeur JSON du type suivant : 

   ```json
   {"accessToken": "TTTTT", "expiredDate": "DDDDD", "refreshToken": "RRRRR"}
   ```

10. Avec chacun des SSO OIDC, se connecter avec un compte du SSO différent de la fois précédente et reprendre toutes les vérifications à partir du 4. mais au niveau de la double mire aller dans la partie _« J’ai déjà un compte Pix »_ et cliquer sur le bouton _« Je me connecte »_

11. Modifier la variable d'environnement précédemment définie comme suit : 

    ```shell
    FWB_CLAIMS_TO_STORE=family_name, given_name, employeeNumber, population
    ```

12. Avec chacun des SSO OIDC, se connecter avec le compte du SSO précédemment utilisé et reprendre toutes les vérifications à partir du 3. La seule chose qui changer dans le déroulé est que la double mire n'est plus affichée car le compte SSO est déjà associé à un compte Pix. Enfin constater que l'`authenticationComplement` pour le compte Pix associé à un compte SSO FWB a été « modifié-écrasé » avec une nouvelle valeur contenant la propriété `population` : 

    ```sql
    select * from "authentication-methods" where "identityProvider"='FWB';
    ```

    avec un `authenticationComplement` qui contient une valeur JSON du type suivant : 
 
    ```json
    { "family_name": "AAAAA", "given_name": "BBBBB", "employeeNumber": "CCCCC", "population": "DDDDD" }
    ```

lego-technix committed:
feat(api): make create-oidc-user, reconcile-oidc-user, authenticate-oidc store claimsToStore in authenticationComplement

except for POLE_EMPLOI

lego-technix committed:
fix(api): fix wrong array copy for OidcAuthenticationService requiredProperties

affecting OidcAuthenticationService DEFAULT_REQUIRED_PROPERTIES.

lego-technix committed:
refactor(api): rename userInfoContent→userInfo

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr7392.review.pix.fr)
  * [App (.org)](https://app-pr7392.review.pix.org)
  * [Orga (.fr)](https://orga-pr7392.review.pix.fr)
  * [Orga (.org)](https://orga-pr7392.review.pix.org)
  * [Certif (.fr)](https://certif-pr7392.review.pix.fr)
  * [Certif (.org)](https://certif-pr7392.review.pix.org)
  * [1D](https://1d-pr7392.review.pix.fr)
  * [Admin](https://admin-pr7392.review.pix.fr)
  * [API](https://api-pr7392.review.pix.fr/api/)
  * [Audit Logger](https://pix-audit-logger-review-pr7392.osc-fr1.scalingo.io/api/)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr7392/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr7392/environment)
  * [scalingo audit-logger](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr7392/environment)

igorissen commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -191,103 +202,51 @@ class OidcAuthenticationService {
+
       const authenticationMethod = new AuthenticationMethod({
         identityProvider: this.identityProvider,
+        ...(claimsToStoreWithValues && { authenticationComplement: { ...claimsToStoreWithValues } }),
```
```suggestion
      let claimsToStoreWithValues = {};
      if (this.claimsToStore) {
        claimsToStoreWithValues = Object.fromEntries(
          Object.entries(userInfo).filter(([key, _value]) => this.claimsToStore.includes(key)),
        );
      }

      const authenticationMethod = new AuthenticationMethod({
        identityProvider: this.identityProvider,
        ...(claimsToStoreWithValues && { authenticationComplement: { ...claimsToStoreWithValues } }),
```
Si la variable `claimsToStoreWithValues` n'est pas initialisée on risque d'avoir l'erreur suivante à la ligne 249. 

![Screenshot 2023-11-07 at 09 42 31](https://github.com/1024pix/pix/assets/6919604/b2963747-cfc6-4dc5-ac8f-87d7245850ed)

igorissen commented on api/lib/domain/services/authentication/oidc-authentication-service.js:
```diff
@@ -22,11 +22,15 @@ import { temporaryStorage } from '../../../infrastructure/temporary-storage/inde
 class OidcAuthenticationService {
   #isReady = false;
 
+  requiredClaims = Array.from(DEFAULT_REQUIRED_CLAIMS);
```
Si les attributs `requiredClaims` et `requiredProperties` ne sont utilisés qu'en interne, on pourrait les passer en attribut privé.

lego-technix replied:
Voilà, `requiredClaims` a été passée en privé avec un fixup commit. Pour `requiredProperties` c'est une variable locale au constructeur et donc ce n'est pas un attribut de la classe.

EmmanuelleBonnemay reviewed (approved):
Testé fonctionnellement avec succès sur Firefox.

igorissen reviewed (approved):
<img src="https://media2.giphy.com/media/ClqoZCSN36ECXUAuj6/giphy.gif"/>

lego-technix committed:
fix(api): fix databaseBuilder wrt to generic OIDC providers and leverage withPoleEmploiAsIdentityProvider

lego-technix committed:
fix(api): add missing PAYSDELALOIRE cases in validations

lego-technix committed:
refactor(api): use a dedicated PoleEmploiOidcAuthenticationComplement for POLE_EMPLOI OIDC Provider

lego-technix committed:
refactor(api): then reintroduce OidcAuthenticationComplement for generic OIDC Providers

lego-technix committed:
test(api): split contexts into generic OIDC Providers and POLE_EMPLOI OIDC Provider

lego-technix committed:
refactor(api): move down createAuthenticationComplement

as it is called after createUserAccount

lego-technix committed:
feat(api): write the authenticationComplement in DB at every authentication

This way the authenticationComplement is always synchronized with the current
state of the user, especially when there is no more claimsToStore configured.
But there is a useless overload when the authenticationComplement has not
changed since the last time. A performance improvement would be to write the
authenticationMethod in DB only if the authenticationComplement has changed, but
this will be done in another PR if performance problems arise.

lego-technix committed:
feat(api): make possible for generic OIDC Providers to have an optional authenticationComplement

igorissen reviewed (approved):
<img src="https://media1.giphy.com/media/NytMLKyiaIh6VH9SPm/giphy.gif"/>

EmmanuelleBonnemay reviewed (approved):
Testé fonctionnellement avec succès sur Firefox et Chromium.
